newIDE/app/src/MainFrame/index.js

```diff
@@ -351,6 +351,7 @@ const MainFrame = (props: Props) => {
   const [isProjectOpening, setIsProjectOpening] = React.useState<boolean>(
     false
   );
+  const [isProjectClosed, setIsProjectClosed] = React.useState<boolean>(false);
```

Comment:
I'm not sure if it's a good idea to have boolean for states.

Reply:
Having boolean for states is ok usually. 
Here I'm unsure about the meaning of this one:
- Are you able to do the same by just looking at the `project`? (as in `if (!project)` being the same as `isProjectClosed` ?
- If not, then the name should probably be different, like the boolean just above `isProjectClosing` ? Then you can set it to true at the beginning of the  `closeProject` function and set it to false at the end with a big `try` / `catch` / `finally` <- in the finally ?

Reply:
> * Are you able to do the same by just looking at the `state.currentProject`? (as in `if (!currentProject)` being the same as `isProjectClosed` ?
> * If not, then the name should probably be different, like the boolean just above `isProjectClosing` ? Then you can set it to true at the beginning of the  `closeProject` function and set it to false at the end with a big `try` / `catch` / `finally` <- in the finally ?

I wanted to do `isProjectClosing` at 1st but it seems complicated to know when a project can be considered closed:
- change the state to close tabs
- tabs detect a change of focus when they are closed and send an event
- the mainframe handle this event and choose to reload extension or not

What if `isProjectClosing` is set to false too early and events still refresh extensions when they shouldn't?
It seems risky.

Maybe `isProjectClosingOrClosed` to be more precised?

Reply:
I agree that we should aim for not having this state if we can, by looking at `currentProject`. If a project is not opened/closed, then `currentProject` should be null.

Reply:
Splitting the setState in 2 is not a viable solution?

Reply:
I see three ideas:
- It's bad to start something when a component is unmounted. I remember this creating issues in the instances property. If you're unmounted, you probably don't want to call your parent, because maybe your parent will try to use a stale variable. In other words: **remove the componentWillUnmount**. Check how to still ensure extensions are reloaded if I do a change in an extension and just close the tab.
- We can consider it's the parent fault to try to do a job on something that was just set to null, but because they are staled they did not. It means the function where you added "isProjectClosed" should use `currentProjectRef`. Because this ref will always stay the same, and the variable inside is always up to date with the latest state.
  **I'm not sure if this will work, but try this**. **Log the value of `currentProjectRef.current`** and log when the `null` is set for currentProject and when it's destroyed. Finally, **log** when something is written by the extension loader.
  I think this might not work because if `currentProjectRef.current` is not `null`, then you still have the same issue: you're asking the extension loader to do a job on something that will be set to null and destroyed from memory very soon (at the next "tick" of the JS event loop).
- We can consider it's the responsibility of the project closing to WAIT FOR EXTENSIONS to be written. Use `await eventsFunctionsExtensionsState.ensureLoadFinished()`.

  In other words, this mean: "you asked the extension loader to do a job on a project, so you might as well wait for it to be done before trying to destroy the thing it is using! ;)"
  
Ideally, we would do all of these:
- avoid if we can to call things on unmounting. 
- even if we do or don't, check if the `currentProject` and `currentProjectRef.current` are different in the callbacks. I think they will be different. 
- even if we do or don't do this, let's still wait for `eventsFunctionsExtensionsState.ensureLoadFinished` because WHO KNOWS, maybe something has been triggering a extension loading, but somehow the hard drive took or network took 10 seconds to write a file, so a promise will resolve after the project is destroyed. So we must wait for this to finish.

Reply:
Let's name this state `isProjectClosedSoAvoidReloadingExtensions`.
It's a terrible, long name, and it's perfect because it should be a hint that if someone tries to use it, it's a terrible idea that was done to show that this should be removed as soon as we can.